File: Stack/Python/Stack.py
# -*- coding: utf-8 -*-

class Stack:
    """ An implementation of a stack """
    MAX_SIZE = 8
    # Initialise list
    stack_ = [""] * MAX_SIZE
    top = 0
 
    def __init__(self, items=None):
        """ Init, take optional argument items """
        # Check the user has specified a list of items
        if(items != None):
            # Check the list of items is not too long
            if(len(items) > self.MAX_SIZE):
                raise IndexError("The list has more than the max number of items in it!")
            else:
                # Put the items in the stack
                for i in range(0, len(items)):
                    self.stack_[i] = items[i]
                # The next available slot will be the same as the length
                self.top = len(items)
        else:
            self.top = 0

    def push(self, item):
        """ Push a value onto the stack """
        if(not self.isFull()):
            self.stack_[self.top] = item
            self.top += 1
        else:
            raise IndexError("The stack is full")
    
    def pop(self):
        """ Pop an item off the stack """
        if(not self.isEmpty()):
            item = self.stack_[self.top - 1]
            self.top -= 1
            return item
        else:
            raise IndexError("Stack is empty, nothing to pop")

    def peek(self):
        """ Peek top item without popping it """
        if(not self.isEmpty()):
            return self.stack_[self.top - 1]
        else:
            raise IndexError("Stack is empty, nothing to peek")

    def clear(self):
        """ Clear the stack """
        top = 0

    def __len__(self):
        """ Hook method for returning size of stack using len(stack) """
        return self.top
    
    def isEmpty(self):
        """ Bool, returns true or false based on whether the stack is empty or not """
        if(self.top == 0):
            return True
        else:
            return False

    def isFull(self):
        """ Bool, returns true or false based on whether the stack is full or not """
        if(self.top == self.MAX_SIZE):
            return True
        else:
            return False

""" Main - Tests """
if(__name__ == "__main__"):
    # Init and add to stack
    s = Stack(["item 1", "item 2", "item 3", "item 4"])
    s.push("item 5")
    s.push("item 6")
    s.push("item 7")
    s.push("item 8")
    # Check if full (yes)
    print("Is full: " + str(s.isFull()))
    # peek then pop
    print("Peek: " + s.peek())
    print("Pop: " + s.pop())
    print("Len: " + str(len(s)))

File: Stack/C++/stack.cpp
#include <iostream>
#include <vector>
#include "stack.h"

using namespace std;

template <class T>
Stack<T>::Stack<T>(vector<T> items = NULL){
    /* Initialize stack, taking optional list of items to place in first */

	// Initialize variables
	MAX_SIZE = 8;
	stack_[MAX_SIZE];
    // Check a list of items exists first
    if(items != NULL){
        // Check length
        if(items.size() > MAX_SIZE){
			throw out_of_range("The list has more than the max number of items in it");
        }
		else {
			// Put items in the stack
			for (int i = 0; i < MAX_SIZE; i++) {
				stack_[i] = items[i]
			}
			top = items.size();
		}
    }
	else {
		top = 0;
	}
}

template <class T>
void Stack<T>::Push(T item){
	/* Push an item onto the stack */
	if (isFull()) {
		throw out_of_range("The stack is full");
	}
	else {
		stack_[top] = item;
		top++;
	}
}

template <class T>
T Stack<T>::pop(){
	/* Pop an item off the stack */
	if(!isEmpty()) {
		T item = stack_[top - 1];
		top--;
		return item;
	}
	else {
		throw out_of_range("Stack is empty, nothing to pop");
	}
}

template <class T>
T Stack<T>::peek(){
	/* Peek top item without popping it */
	if (!isEmpty()) {
		return stack_[top - 1];
	}
	else {
		throw out_of_range("Stack is empty, nothing to peek");
	}
}

template <class T>
void Stack<T>::clear(){
	/* Clear stack */
	top = 0;
}

template <class T>
int Stack<T>::getLength(){
	/* Return size of stack */
	return top;
}

template <class T>
bool Stack<T>::isEmpty(){
	/* Returns true or false based on whether the stack is empty or not */
	if (top == 0) {
		return true;
	}
	else {
		return false;
	}
}

template <class T>
bool Stack<T>::isFull(){
	/* Returns true or false based on whether the stack is full or not */
	if (top == MAX_SIZE) {
		return true;
	}
	else {
		return false;
	}
}